### ISSUE TITLE
Sparse updates to logits in distributions.Categorical

### ISSUE THREAD

**Issue by Ajk4** [pytorch/pytorch#62473]
## :rocket: Feature

Currently in distributions.Categorical if we backward through logprobs we touch every single paramer.
However by .detach()-ing normalization from graph we can make this update sparse and save a lot of memory/time when dealing
with huge categorical variables.

## Motivation

I'm dealing with huge Categorical variables and I'm trying to make it fit in my memory. In my use-case Categorical distributions uses free-variables for its logits and it's not a function of other parameters. It makes it preferable for me to use sparse-updates to save memory.

## Pitch

By .detach()-ing normalization of logits we can make updates small and sparse. Net-probability change stays the same.

## Alternatives

I understand that dense update might be preferred in some cases. Perhaps we could control it by parameter?

## Additional context

AFAIK .detach()-ing or not normalization factor does not make a difference from probability point of vied. Net-change of probability will be the same. Increasing one logit locally is equivalent to changing decreasing all other logits by 1/N.

![image](https://user-images.githubusercontent.com/2959412/127632392-ac38ef32-5bba-4ecd-8b56-bf856a522aef.png)

**Comment by fritzo** [comment]
@Ajk4 this seems reasonable to me, but a couple caveats:
1. Many PyTorch users have been confused by the gradients to the constrained `probs` parameter. I forget the issue numbers, but there have been many issues opened claiming at first to find a bug, then eventually conceding that in any unconstrained parametrization of `probs` the effective gradients are correct. Anyway if you do decide to submit a PR, it would probably save you some cc'd questions in the future if you explain in detail what's going on in the docstring.
2. I have [anecdotally](https://github.com/pyro-ppl/pyro/pull/2710#discussion_r537900148) seen better numerical stability in the undetached version `logits - logits.logsumexp(-1, True)` than in the detached version `logits - logits.detach().logsumexp(-1, True)`
3. I think a nice way to do this would be to provide an `is_sparse` kwarg, say
  ```diff
  - def __init__(self, probs=None, logits=None, validate_args=None):
  + def __init__(self, probs=None, logits=None, validate_args=None, *, is_sparse=False):
  ```
  (I think we should eventually *-guard the `validate_args` keyword as well, but I defer to @neerajprad for strategy for this breaking change)

**Comment by albanD** [comment]
Note that we have a similar `sparse` argument for the embedding layer (https://pytorch.org/docs/stable/generated/torch.nn.Embedding.html?highlight=embedding#torch.nn.Embedding) that we might want to mimick here.
